vortex-array/src/stats.rs

```diff
@@ -24,150 +26,295 @@ pub enum Stat {
     NullCount,
 }
 
+impl Display for Stat {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Stat::BitWidthFreq => write!(f, "bit_width_frequency"),
+            Stat::TrailingZeroFreq => write!(f, "trailing_zero_frequency"),
+            Stat::IsConstant => write!(f, "is_constant"),
+            Stat::IsSorted => write!(f, "is_sorted"),
+            Stat::IsStrictSorted => write!(f, "is_strict_sorted"),
+            Stat::Max => write!(f, "max"),
+            Stat::Min => write!(f, "min"),
+            Stat::RunCount => write!(f, "run_count"),
+            Stat::TrueCount => write!(f, "true_count"),
+            Stat::NullCount => write!(f, "null_count"),
+        }
+    }
+}
+
+pub trait ArrayStatistics {
+    fn statistics(&self) -> &dyn Statistics;
+}
+
+pub trait Statistics {
+    /// Returns the value of the statistic only if it's present
+    fn get(&self, stat: Stat) -> Option<Scalar>;
+
+    /// Get all existing statistics
+    fn get_all(&self) -> StatsSet;
+
+    fn set(&self, stat: Stat, value: Scalar);
+
+    fn set_many(&self, other: &dyn Statistics, stats: &[Stat]);
+
+    /// Computes the value of the stat if it's not present
+    fn compute(&self, stat: Stat) -> Option<Scalar>;
+}
+
+pub trait OwnedStats {
+    fn stats_set(&self) -> &RwLock<StatsSet>;
+}
+
+pub trait StatsCompute {
+    fn compute(&self, stat: Stat) -> VortexResult<StatsSet>;
+}
+
+impl<T: OwnedStats + StatsCompute> Statistics for T {
```

Comment:
Since we have this impl we could also add blanket impl for `ArrayStatistics` but I wasn't sure if we can figure out a way to make it not conflict. This one might also have to change but for now everything has `OwnedStats`